# Introducción a las APIs

## 1. Introducción a las APIs
Las APIs, o Interfaces de Programación de Aplicaciones, son un conjunto esencial de herramientas y protocolos en el mundo de la programación y el desarrollo de software.

### ¿Qué son las APIs?

Una API es una interfaz que permite a diferentes sistemas de software interactuar entre sí. Actúa como un intermediario, permitiendo que una aplicación envíe y reciba datos de otra aplicación o plataforma. 

![API](../img/API.png)

Las APIs definen las reglas que los programadores deben seguir para interactuar con un servicio de software, incluyendo cómo realizar solicitudes, los formatos de datos necesarios, y los protocolos de comunicación a utilizar.

### Importancia de las APIs en el Desarrollo Moderno

![API_CAR](../img/car_api.png)


1. **Integración y Conectividad**: En el contexto de la creciente interconexión del mundo digital, las Interfaces de Programación de Aplicaciones (APIs) desempeñan un papel fundamental en la vinculación de sistemas heterogéneos, facilitando la compartición de datos y funcionalidades entre ellos. Esta capacidad es crucial para el desarrollo de experiencias de usuario cohesivas y eficientes, permitiendo una integración fluida entre diversas plataformas y aplicaciones.

2. **Innovación y Agilidad**: Las APIs son catalizadores de la innovación tecnológica, ya que ofrecen a los desarrolladores la posibilidad de utilizar funcionalidades ya establecidas en lugar de desarrollar soluciones desde cero. Este enfoque no solo agiliza significativamente el ciclo de desarrollo, sino que también proporciona a las organizaciones la flexibilidad necesaria para adaptarse rápidamente a las cambiantes demandas y tendencias del mercado.

3. **Economía de las APIs**: En la economía digital actual, numerosas empresas han adoptado modelos de negocios centrados en el uso de APIs. Al ofrecer sus servicios a través de estas interfaces, las organizaciones facilitan su utilización por parte de otros desarrolladores, dando lugar a la creación de un mercado dinámico caracterizado por oportunidades ampliadas de colaboración y generación de ingresos.

4. **Escalabilidad**: Una de las ventajas más destacadas de las APIs es su capacidad para facilitar la escalabilidad de productos y servicios. Mediante la segregación de funcionalidades en distintas APIs, las empresas pueden expandir sus servicios de manera más efectiva y eficiente, adaptándose al crecimiento de la demanda y a la expansión del negocio.

5. **Automatización**: El uso de APIs es un factor clave en la automatización de procesos empresariales. A través de su implementación, las empresas pueden automatizar tareas repetitivas, integrar sistemas dispares y, en consecuencia, optimizar la eficiencia operativa, reduciendo el tiempo y los recursos invertidos en tareas manuales.

6. **Personalización**: Las APIs ofrecen un medio eficaz para personalizar la experiencia del usuario final. Mediante la integración y combinación de diversos servicios y datos, los desarrolladores tienen la capacidad de crear aplicaciones que se ajustan específicamente a las necesidades y preferencias individuales de los usuarios, mejorando así la relevancia y el valor de las soluciones ofrecidas.

Las APIs son fundamentales en la era digital actual, permitiendo una mayor colaboración, innovación y eficiencia en el desarrollo y la integración de software. Son herramientas esenciales para cualquier desarrollador que busque crear soluciones modernas y conectadas.

## 2. ¿Cómo Funcionan las APIs?


El funcionamiento de una API se puede entender a través del modelo de solicitud-respuesta. Cuando una aplicación desea acceder a recursos o realizar una operación en otra aplicación, envía una solicitud a través de la API. Esta solicitud incluye información específica sobre lo que se requiere y cómo se debe entregar. La API, actuando como intermediaria, procesa esta solicitud, la envía al sistema adecuado y, luego, devuelve la respuesta al solicitante.

![API_RESPONSE](../img/api_request-response_flow.png)


En ese sentido, entendiendo a las APIs como un conjunto de reglas que definen cómo se debe interactuar con un servicio de software, es importante tener en cuenta sus diferentes componentes:

1. Interfaz de Comunicación: Define los métodos de comunicación que se pueden utilizar para acceder a un servicio de software. Estos métodos incluyen solicitudes HTTP, llamadas a procedimientos remotos (RPC), y mensajes de intercambio de información (MIME).

### 1. **Definición de Contratos**:
   - **Especificación de Interfaz**: Las APIs definen una interfaz o contrato que establece cómo deben interactuar las aplicaciones. Esto incluye detalles como qué tipos de solicitudes se pueden hacer y qué respuestas se pueden esperar.
   - **Documentación Clave**: La documentación de la API es esencial. Debe incluir información detallada sobre los métodos, parámetros, formatos de datos y ejemplos de uso.

### 2. Autenticación y Seguridad:

**Autenticación**: La autenticación es el proceso de verificar la identidad de un usuario. Las APIs suelen utilizar tokens de acceso para autenticar a los usuarios. Estos tokens se generan cuando un usuario se registra en una aplicación y se utilizan para identificar al usuario en las solicitudes posteriores.

   Las autorizaciones de cada usuario en la API puede variar en complejidad y puede implementarse de diferentes formas, pero en general, involucran la autenticación (verificación de la identidad) y la concesión de permisos específicos (autorización) para acceder a recursos o realizar ciertas acciones en la API.   

   En el contexto de OAuth (Open Authorization), que es un protocolo comúnmente utilizado para autorizar aplicaciones de terceros a acceder a recursos en nombre de un usuario, las autorizaciones se basan en la emisión de tokens de acceso temporales y la definición de ámbitos (scopes) que especifican qué recursos y acciones están permitidos.   
   
   Los conceptos clave relacionados con las autorizaciones en las APIs se resumen a continuación:

   
   | Concepto                  | Descripción                                                                                                         |
|---------------------------|---------------------------------------------------------------------------------------------------------------------|
| Claves de API (API Keys)  | Credenciales proporcionadas por el proveedor de la API para autenticar y acceder a la API. Pueden ser públicas o privadas.  |
| OAuth (Open Authorization) | Protocolo estándar de autorización que permite que una aplicación acceda a recursos en nombre de un usuario sin compartir contraseñas. |
| Tokens de Acceso          | Credenciales temporales utilizadas en lugar de contraseñas o claves de API para acceder a recursos protegidos. Tienen un tiempo de vida limitado. |
| Flujo de Autorización OAuth2 | Proceso que incluye solicitud de autorización, inicio de sesión del usuario en la plataforma de autenticación, obtención de un código de autorización y el intercambio de este código por un token de acceso. |
| Scopes (Ámbitos)          | Permisos específicos otorgados por el usuario a una aplicación durante el proceso de autorización. Define qué recursos puede acceder la aplicación. |
| Seguridad                 | La importancia de proteger las claves de API y los tokens de acceso para evitar accesos no autorizados. Nunca deben ser compartidos públicamente. |
| Límites de Tasa (Rate Limits) | Restricciones en la cantidad de solicitudes que se pueden realizar en un período de tiempo determinado para evitar un uso excesivo o abusivo de la API. |

Estos conceptos son fundamentales para comprender cómo funcionan las autorizaciones en las APIs y cómo se protege la seguridad y la privacidad de los datos de los usuarios.


**Seguridad**: La seguridad en las APIs es un componente esencial del desarrollo de aplicaciones y servicios en línea. Las buenas prácticas de seguridad y la elección adecuada de protocolos de seguridad son fundamentales para proteger los recursos y datos expuestos a través de una API y para garantizar la confianza de los usuarios en la plataforma. La implementación cuidadosa de estas medidas contribuye a mitigar riesgos y prevenir amenazas a la seguridad.

Tener en cuenta: 

**Buenas Prácticas de Seguridad en APIs:**


1. **Autenticación y Autorización**: Establecer un sistema sólido de autenticación y autorización que verifique la identidad de los usuarios o aplicaciones y defina qué recursos pueden acceder y qué acciones pueden realizar.

2. **Validación de Datos de Entrada y Salida**: Validar y sanitizar todos los datos de entrada y salida de la API para prevenir ataques comunes, como inyección de SQL y ataques de cross-site scripting (XSS).

3. **Protección contra Ataques**: Implementar medidas de seguridad contra ataques conocidos, como cross-site request forgery (CSRF) y protección contra inyección de código.

4. **Gestión de Tokens y Claves de API**: Manejar las claves de API y tokens de acceso de manera segura, evitando su exposición pública y estableciendo políticas de tiempo de vida limitado.

5. **Límites de Tasa (Rate Limits)**: Establecer límites de tasa para las solicitudes a la API para evitar un uso excesivo o abusivo de los recursos de la API.

**Protocolos de Seguridad:**


1. **OAuth (Open Authorization)**: OAuth es un protocolo ampliamente utilizado para autorizar aplicaciones de terceros a acceder a recursos en nombre de un usuario. Permite que los usuarios otorguen permisos específicos a aplicaciones y controlen su acceso.

2. **HTTPS (TLS/SSL)**: Utilizar HTTPS (HyperText Transfer Protocol Secure) con cifrado SSL/TLS es esencial para proteger la comunicación entre el cliente y la API. Evita la interceptación de datos y garantiza la confidencialidad.

3. **JWT (JSON Web Tokens)**: Los JWT son un formato compacto y seguro para transmitir información entre partes de manera confiable. Se utilizan comúnmente para la autenticación y la autorización en APIs.

4. **SAML (Security Assertion Markup Language)**: SAML es un estándar de autenticación y autorización basado en XML utilizado en aplicaciones empresariales para el intercambio de información de autenticación.


### 3. **Métodos y Endpoints**:
   - **Métodos HTTP**: Las API web suelen utilizar métodos HTTP estándar, como GET (para recuperar datos), POST (para crear nuevos datos), PUT (para actualizar datos existentes) y DELETE (para eliminar datos).

   ![image.png](../img/GETPOSTPUT.png)


   - **Endpoints**: Cada método se vincula a una URL específica llamada "endpoint". Por ejemplo, un endpoint para obtener información sobre un usuario podría ser `https://api.example.com/users/123`.

   Ejemplos:  

In [8]:
import requests

# URL de la API de REST Countries
url = "https://restcountries.com/v3.1/name/"

# Nombre del país que deseas consultar
pais = "Colombia"

# Realiza la solicitud GET a la API
response = requests.get(url + pais)
print(response._content)

b'[{"name":{"common":"Colombia","official":"Republic of Colombia","nativeName":{"spa":{"official":"Rep\xc3\xbablica de Colombia","common":"Colombia"}}},"tld":[".co"],"cca2":"CO","ccn3":"170","cca3":"COL","cioc":"COL","independent":true,"status":"officially-assigned","unMember":true,"currencies":{"COP":{"name":"Colombian peso","symbol":"$"}},"idd":{"root":"+5","suffixes":["7"]},"capital":["Bogot\xc3\xa1"],"altSpellings":["CO","Republic of Colombia","Rep\xc3\xbablica de Colombia"],"region":"Americas","subregion":"South America","languages":{"spa":"Spanish"},"translations":{"ara":{"official":"\xd8\xac\xd9\x85\xd9\x87\xd9\x88\xd8\xb1\xd9\x8a\xd8\xa9 \xd9\x83\xd9\x88\xd9\x84\xd9\x88\xd9\x85\xd8\xa8\xd9\x8a\xd8\xa7","common":"\xd9\x83\xd9\x88\xd9\x84\xd9\x88\xd9\x85\xd8\xa8\xd9\x8a\xd8\xa7"},"bre":{"official":"Republik Kolombia","common":"Kolombia"},"ces":{"official":"Kolumbijsk\xc3\xa1 republika","common":"Kolumbie"},"cym":{"official":"Gweriniaeth Colombia","common":"Colombia"},"deu":{"offi

In [9]:

# Verifica si la solicitud fue exitosa (código de estado 200)
if response.status_code == 200:
    # Convierte la respuesta JSON en un diccionario de Python
    data = response.json()

    # Accede a la información del país
    nombre = data[0]["name"]["common"]
    capital = data[0]["capital"][0]
    poblacion = data[0]["population"]
    region = data[0]["region"]
    subregion = data[0]["subregion"]

    # Imprime la información del país
    print(f"Nombre: {nombre}")
    print(f"Capital: {capital}")
    print(f"Población: {poblacion}")
    print(f"Región: {region}")
    print(f"Subregión: {subregion}")
else:
    print("No se pudo obtener información del país.")


Nombre: Colombia
Capital: Bogotá
Población: 50882884
Región: Americas
Subregión: South America


In [2]:
### Usando POST

import requests
import json

# URL de la API JSONPlaceholder para crear un nuevo usuario
url = "https://jsonplaceholder.typicode.com/users"

# Datos del nuevo usuario
nuevo_usuario = {
    "name": "Juanito Alimaña",
    "username": "jalimana",
    "email": "jalimana@example.com"
}

# Convierte los datos a formato JSON
data = json.dumps(nuevo_usuario)

# Cabeceras de la solicitud
headers = {
    "Content-Type": "application/json; charset=UTF-8"
}

# Realiza la solicitud POST para crear el nuevo usuario
response = requests.post(url, data=data, headers=headers)

# Verifica si la solicitud fue exitosa (código de estado 201 para creación exitosa)
if response.status_code == 201:
    # Convierte la respuesta JSON en un diccionario de Python
    nuevo_usuario_creado = response.json()

    # Imprime la respuesta que incluirá el ID asignado por el servidor
    print("Nuevo usuario creado:")
    print("ID:", nuevo_usuario_creado["id"])
    print("Nombre:", nuevo_usuario_creado["name"])
    print("Nombre de usuario:", nuevo_usuario_creado["username"])
    print("Email:", nuevo_usuario_creado["email"])
else:
    print("No se pudo crear el nuevo usuario. Código de estado:", response.status_code)


Nuevo usuario creado:
ID: 11
Nombre: Juanito Alimaña
Nombre de usuario: jalimana
Email: jalimana@example.com


In [3]:
import requests

# URL de la API JSONPlaceholder para obtener información de usuarios
url = "https://jsonplaceholder.typicode.com/users"

# ID del usuario que deseas verificar (debe coincidir con el ID del usuario creado anteriormente)
id_usuario_creado = 11  # Reemplaza esto con el ID del usuario creado

# Realiza una solicitud GET para obtener la información del usuario
response = requests.get(f"{url}/{id_usuario_creado}")

# Verifica si la solicitud fue exitosa (código de estado 200)
if response.status_code == 200:
    # Convierte la respuesta JSON en un diccionario de Python
    usuario_obtenido = response.json()

    # Imprime la información del usuario
    print("Información del usuario obtenida:")
    print("ID:", usuario_obtenido["id"])
    print("Nombre:", usuario_obtenido["name"])
    print("Nombre de usuario:", usuario_obtenido["username"])
    print("Email:", usuario_obtenido["email"])
else:
    print("No se pudo obtener la información del usuario. Código de estado:", response.status_code)


No se pudo obtener la información del usuario. Código de estado: 404


### 3. Parámetros:

Los parámetros desempeñan un papel fundamental en el funcionamiento de las APIs, ya que permiten a los clientes personalizar sus solicitudes y obtener resultados específicos. Aquí se describen dos aspectos importantes relacionados con los parámetros en las APIs:

#### Parámetros de Solicitud:

Los parámetros de solicitud son valores que los clientes envían junto con una solicitud a una API para influir en el comportamiento de la solicitud o personalizar la respuesta. Estos parámetros se incluyen típicamente en la URL de la solicitud o en el cuerpo de la solicitud, según el método HTTP utilizado (GET o POST, respectivamente). Algunos ejemplos comunes de parámetros de solicitud incluyen:

- **Parámetros de consulta (Query Parameters)**: Estos parámetros se incluyen en la URL y se utilizan para filtrar, ordenar o paginar resultados. Por ejemplo, al consultar una lista de productos, se podría especificar un parámetro de consulta para obtener solo los productos de una categoría específica.

  Ejemplo de URL con parámetros de consulta:
  ```
  https://api.example.com/products?category=electronics&page=2&limit=10
  ```

- **Parámetros en el cuerpo de la solicitud (Request Body)**: En solicitudes POST o PUT, los parámetros pueden enviarse en el cuerpo de la solicitud en formato JSON o XML. Esto se utiliza para proporcionar datos estructurados al servidor, como al crear o actualizar recursos.

  Ejemplo de cuerpo de solicitud JSON con parámetros:
  ```json
  {
      "name": "Nuevo Producto",
      "price": 29.99,
      "category": "Electrónicos"
  }
  ```

#### Validación de Parámetros:

La validación de parámetros es un paso crítico en el procesamiento de solicitudes de API. Consiste en asegurarse de que los parámetros proporcionados sean correctos y estén dentro de los límites permitidos antes de procesar la solicitud. La validación de parámetros es esencial por varias razones:

- **Seguridad**: La validación evita posibles ataques, como la inyección de SQL o el cross-site scripting (XSS), al garantizar que los datos ingresados cumplan con las expectativas y no contengan contenido malicioso.

- **Integridad de Datos**: Ayuda a mantener la integridad de los datos al verificar que los parámetros estén en el formato correcto y cumplan con las restricciones definidas.

- **Eficiencia**: La validación temprana de parámetros puede ahorrar tiempo y recursos al evitar el procesamiento innecesario de solicitudes con parámetros incorrectos.

- **Experiencia del Usuario**: Proporciona una mejor experiencia de usuario al informar a los clientes sobre errores de validación de manera clara y comprensible.

La validación de parámetros puede incluir la verificación de tipos de datos, la comprobación de rangos válidos, la autenticación de tokens de acceso y la validación de campos requeridos, entre otros aspectos.

### 4. **Respuestas**:

Las respuestas de una API son la manera en que la API comunica el resultado de una solicitud realizada por un cliente. Aquí se explican dos componentes clave de las respuestas de una API:

#### Códigos de Estado HTTP:

Los códigos de estado HTTP son números de tres dígitos que indican el resultado de una solicitud HTTP. Estos códigos proporcionan información sobre si la solicitud se completó con éxito, si hubo un error o si se requiere alguna acción adicional por parte del cliente. Algunos ejemplos comunes de códigos de estado HTTP incluyen:

| Código de Estado HTTP | Descripción |
|-----------------------|-------------|
| 200 OK                | Indica que la solicitud se completó con éxito y se devuelven los datos solicitados en el cuerpo de la respuesta. |
| 201 Created           | Se utiliza cuando se crea un nuevo recurso en el servidor como resultado de la solicitud (por ejemplo, al crear un nuevo usuario). |
| 204 No Content        | Indica que la solicitud se completó con éxito, pero no hay datos adicionales que devolver en el cuerpo de la respuesta. |
| 400 Bad Request       | Se utiliza cuando la solicitud del cliente contiene datos incorrectos o no válidos. |
| 401 Unauthorized      | Indica que la solicitud requiere autenticación y el cliente no proporcionó credenciales válidas. |
| 404 Not Found         | Se utiliza cuando el recurso solicitado no existe en el servidor. |
| 500 Internal Server Error | Indica un error interno en el servidor que impide que la solicitud se complete con éxito. |

Estos códigos de estado HTTP son estándares y se utilizan para comunicar el resultado de una solicitud de manera clara y consistente en toda la web. Cada código tiene un significado específico y ayuda a los clientes a comprender lo que sucedió con su solicitud.

#### Datos de Respuesta:

Junto con el código de estado HTTP, una API devuelve datos relevantes en el cuerpo de la respuesta. Estos datos suelen estar en formato JSON, XML u otro formato estructurado y contienen la información solicitada por el cliente o mensajes de error detallados. Los datos de respuesta pueden variar según la solicitud y el tipo de recurso solicitado. Algunos ejemplos de datos de respuesta incluyen:

- **Datos de Recurso**: Si la solicitud es para obtener información sobre un recurso (por ejemplo, un usuario o un producto), los datos de respuesta contendrán los detalles de ese recurso.

  Ejemplo de respuesta JSON con datos de usuario:
  ```json
  {
      "id": 1,
      "name": "John Doe",
      "email": "johndoe@example.com"
  }
  ```

- **Mensajes de Error**: Si se produce un error, los datos de respuesta pueden contener mensajes de error que describen la causa del problema.

  Ejemplo de respuesta JSON con mensaje de error:
  ```json
  {
      "error": "El recurso solicitado no se encontró."
  }
  ```

Los datos de respuesta son fundamentales para que el cliente comprenda el resultado de su solicitud y pueda tomar las acciones apropiadas en función de la información proporcionada.


### 5. Versionado:

El versionado en el contexto de las APIs se refiere a la práctica de mantener múltiples versiones de una API a medida que esta evoluciona con el tiempo. El versionado de la APIse hace para garantizar la retrocompatibilidad y la continuidad de las aplicaciones y servicios que dependen de ella. Aquí se describen los aspectos clave del versionado de la API:

#### Versionado de la API:
El versionado de una API permite a los desarrolladores elegir la versión adecuada y garantiza la continuidad de las aplicaciones existentes a medida que la API evoluciona.


| Aspecto                 | Descripción                                                                                                   | Ejemplo                                      |
|-------------------------|---------------------------------------------------------------------------------------------------------------|----------------------------------------------|
| Múltiples Versiones     | Se crean múltiples versiones de la API a medida que evoluciona, cada una representando un conjunto específico de endpoints y comportamientos. | Versión 1 (v1), Versión 2 (v2), Versión 3 (v3) |
| Nomenclatura de Versiones | Las versiones de la API se denotan en la URL de la solicitud o a través de encabezados personalizados, siguiendo un formato como "v1", "v2", "v3", etc. | /v1/users, /v2/users, /v3/users              |
| Retrocompatibilidad     | Las versiones posteriores de la API deben ser compatibles con las versiones anteriores para que las aplicaciones existentes sigan funcionando. | Si una aplicación utiliza v1, seguirá funcionando con v2. |
| Deprecación             | Las versiones anteriores se degradan gradualmente, alentando a los desarrolladores a migrar a versiones más recientes, pero manteniendo la compatibilidad durante un período antes de ser descontinuadas. | Se deprecia v1, se alienta a los desarrolladores a usar v2, pero v1 sigue funcionando durante 1 año antes de ser descontinuada. |


#### Beneficios del Versionado de la API:

- **Flexibilidad**: El versionado permite a los desarrolladores elegir la versión de la API que mejor se adapte a sus necesidades y mantener aplicaciones existentes en funcionamiento.

- **Control de Cambios**: Facilita la gestión de cambios en la API sin afectar negativamente a las aplicaciones que ya la utilizan.

- **Evolución Continua**: Permite que la API evolucione y se mejore con el tiempo para agregar nuevas características y funcionalidades sin interrupciones importantes.

- **Documentación Clara**: Cada versión de la API puede tener su propia documentación clara y actualizada, lo que facilita a los desarrolladores entender cómo usarla.

